src/num_methods/simulated_annealing.md

```diff
@@ -0,0 +1,198 @@
+
+```cpp
+bool P(double E,double E_next,double T){
+    double prob = (double)rand()/RAND_MAX; // Generate a random number between 0 and 1
+    if(exp(-(E_next-E)/T) > prob) return true;
+    else return false;
+}
+```
+## Code Template
+
+```cpp
+class state{
+    public:
+    state(){
+        // Generate the initial state
+    }
+    state(state& s){
+        // Implement the copy constructor
+    }
+    state next(){
+        state s_next = state(*this);
+
+        // Modify s_next to the neighbouring state
+
+        return s_next;
+    }
+    double E(){
+        // Implement the cost function here
+    };
+};
+
+pair<double,state> simAnneal(){
+    state s = state();
+    state best = state(s);
+    double T = 1000; // Initial temperature
+    double u = 0.99; // decay rate
+    double E = s.E();
+    double E_next;
+    double E_best = E;
+    while (T > 1){
+        state next = s.next();
+        E_next = next.E();
+        if(P(E,E_next,T)){
+            s = next;
+            if(E_next < E_best){
+                best = s;
+                E_best = E_next;
+            }
+        }
+        E = E_next;
+        T *= u;
+    }
+    return {E_best,best};
+}
+
+```
+## How to use:
+Fill in the state class functions as appropriate. If you are trying to find a global maxima and not a minima, ensure that the $E()$ function returns negative of the function you are maximising and finally print out $-E_{best}$. Set the below parameters as per your need.
+
+### Parameters
+- $T$ : Temperature. Set it to a higher value if you want the search to run for a longer time.
+- $u$ : Decay. Decides the rate of cooling. A slower cooling rate (larger value of u) usually gives better results, at the cost of running for a longer time. Ensure $u < 1$. 
+
+The number of iterations the loop will run for is given by the expression
+
+$N =   \lceil -\log_{u}{T} \rceil$ 
+
+Tips for choosing $T$ and $u$ : If there are many local minimas and a wide state space, set $u = 0.999$, for a slow cooling rate, which will allow the algorithm to explore more possibilities. On the other hand, if the state space is narrower, $u = 0.99$ should suffice. If you are not sure, play it safe by setting $u = 0.998$ or higher. Calculate the time complexity of a single iteration of the algorithm, and use this to approximate a value of $N$ which will prevent TLE, then use the below formula to obtain $T$.
+
+$T = u^{-N}$
+
+### Example implementation for TSP
+```cpp
+
+class state{
+    public:
+    vector<pair<int,int>> points;
+
+    state(){
+        {% raw %}points = {{0,0},{2,2},{0,2},{2,0},{0,1},{1,2},{2,1},{1,0}};{% endraw %} 
+    }
+    state(state& s){
+        points = s.points;
+    }
+    state next(){
+        state s_next = state(*this);
+        int a = (rand()%points.size());
```

Comment:
Using `rand()` is [not safe](https://codeforces.com/blog/entry/61587). It may be acceptable in some cases, but in this particular case, it can severely backfire e.g. because on some systems it's always a 16-bit number.